library/aria.c

```diff
@@ -0,0 +1,706 @@
+    0x29, 0xAE, 0x92, 0xD7, 0x84, 0xE9, 0xD2, 0xBA, 0x5D, 0xF3, 0xC5, 0xB0,
+    0xBF, 0xA4, 0x3B, 0x71, 0x44, 0x46, 0x2B, 0xFC, 0xEB, 0x6F, 0xD5, 0xF6,
+    0x14, 0xFE, 0x7C, 0x70, 0x5A, 0x7D, 0xFD, 0x2F, 0x18, 0x83, 0x16, 0xA5,
+    0x91, 0x1F, 0x05, 0x95, 0x74, 0xA9, 0xC1, 0x5B, 0x4A, 0x85, 0x6D, 0x13,
+    0x07, 0x4F, 0x4E, 0x45, 0xB2, 0x0F, 0xC9, 0x1C, 0xA6, 0xBC, 0xEC, 0x73,
+    0x90, 0x7B, 0xCF, 0x59, 0x8F, 0xA1, 0xF9, 0x2D, 0xF2, 0xB1, 0x00, 0x94,
+    0x37, 0x9F, 0xD0, 0x2E, 0x9C, 0x6E, 0x28, 0x3F, 0x80, 0xF0, 0x3D, 0xD3,
+    0x25, 0x8A, 0xB5, 0xE7, 0x42, 0xB3, 0xC7, 0xEA, 0xF7, 0x4C, 0x11, 0x33,
+    0x03, 0xA2, 0xAC, 0x60
+};
+// FO and FE are helpers for key schedule
+
+// r = FO( p, k ) ^ x
+
+static void aria_fo( uint32_t r[4],
+    const uint32_t p[4], const uint32_t k[4], const uint32_t x[4] )
+{
+    uint32_t a, b, c, d;
+    uint32_t t, u, v;
+
+    a = p[0] ^ k[0];
+    b = p[1] ^ k[1];
+    c = p[2] ^ k[2];
+    d = p[3] ^ k[3];
+
+    ARIA_SLA( a, b, c, d, aria_sb1, aria_sb2, aria_is1, aria_is2, t, u, v );
+
+    r[0] = a ^ x[0];
+    r[1] = b ^ x[1];
+    r[2] = c ^ x[2];
+    r[3] = d ^ x[3];
+}
+
+// r = FE( p, k ) ^ x
+
+static void aria_fe(uint32_t r[4],
+    const uint32_t p[4], const uint32_t k[4], const uint32_t x[4] )
+{
+    uint32_t a, b, c, d;
+    uint32_t t, u, v;
+
+    a = p[0] ^ k[0];
+    b = p[1] ^ k[1];
+    c = p[2] ^ k[2];
+    d = p[3] ^ k[3];
+
+    ARIA_SLA( a, b, c, d, aria_is1, aria_is2, aria_sb1, aria_sb2, t, u, v );
+
+    r[0] = a ^ x[0];
+    r[1] = b ^ x[1];
+    r[2] = c ^ x[2];
+    r[3] = d ^ x[3];
+}
+
+// Big endian 128-bit rotation: d = a ^ (b <<< n), used only in key setup.
+// This is relatively slow since our implementation is geared towards
+// little-endian targets and stores state in that order.
+
+static void aria_rot128(uint32_t r[4], const uint32_t a[4],
+                        const uint32_t b[4], int n)
+{
+    int i, j, n1, n2;
+    uint32_t t, u;
+
+    j = (n >> 5) & 3;                       // word offset
+    n1 = n & 0x1F;                          // bit offsets
+    n2 = 32 - n1;                           // n1 should be nonzero!
+    t = ARIA_FLIP1( ARIA_FLIP2( b[j] ) );   // big endian
+    for( i = 0; i < 4; i++ )
+    {
+        j = (j + 1) & 3;                    // get next word, big endian
+        u = ARIA_FLIP1( ARIA_FLIP2( b[j] ) );
+        t <<= n1;                           // rotate
+        if (n2 < 32)                        // intel rotate 32 bits = 0 bits..
+            t |= u >> n2;
+        t = ARIA_FLIP1( ARIA_FLIP2( t ) );  // back to little endian
+        r[i] = a[i] ^ t;                    // store
+        t = u;                              // move to next word
+    }
+}
+
+// Set encryption key
+
+int mbedtls_aria_setkey_enc(mbedtls_aria_context *ctx,
+                            const unsigned char *key, unsigned int keybits)
+{
+    // round constant masks
+    const uint32_t rc[3][4] =
```

Comment:
I thought that the compiler wouldn't store this array in data memory but use the values directly everywhere. But in this case the storage modifier wouldn't make any difference (this is a local variable). What am I missing?

Reply:
I checked the code produced by arm-gcc and (gcc and clang for x86_64) and you're right, the `static` qualifier is useless, either because the compiler generates the same code without it (all with -Os, clang with all -O levels), or because it generates code that is equally acceptable.

So please ignore that particular comment of mine.